brave/brave-core
# Refactoring to migrate hardware and selected accounts prefs under keyrings

Resolves https://github.com/brave/brave-browser/issues/21015

- Fixed coin type save for hardware accounts
- Updated the API `SetDefaultKeyringHardwareAccountName` to `SetHardwareAccountName(string address, string name, CoinType coin)`
- Updated the API `SetSelectedAccount` to `SetSelectedAccount(string address, CoinType coin)`
- Updated the API `RemoveHardwareAccount` to `RemoveHardwareAccount(string address, CoinType coin)`
- Updated the API `SetSelectedAccount` to `SetSelectedAccount(string address, CoinType coin)`
- Selected accounts migrated under keyrings preferences to have separate accounts for each keyring. `GetSelectedAccount` requires `CoinType` as well.
- Moved hardware accounts into keyrings preferences. Migrated all obsolete prefs only to default keyring because we support only ETH hardware accounts at this time. New structure:
```
  "keyrings":
    {
        "default":
        {
            "selected_account": "0x..",
            "account_metas": {},
            "imported_accounts":[],
            "hardware": {
                "3939e7b....55157e81ae7": {
                      "account_metas": {},
                }
            },
        },
        "filecoin": {
            "selected_account": "t1...",
            "hardware": {
                "4939e7b....55157e81ae7": {
                      "account_metas": {},
                }
            },
        },
        "solana": {
            "selected_account": "xxx",
            "hardware": {
                "5939e7b....55157e81ae7": {
                      "account_metas": {},
                }
            },
        }
    }
```



## Submitter Checklist:


## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:

- Import hardware accounts in any browser without this code, beta/stable channels
- Enable either Solana or Filecoin keyring by creation/import accounts for them
- Open accounts page and check hardware accounts are duplicated
- Update to this build
- Check accounts page, duplication disappeared and in the preferences hardware accounts were moved to default keyring.
- Make sure main accounts selection works as expected for different keyrings(ETH/SOL/FIL) because we changed the interface
